File: src/arxiv_copilot/arxiv.py
"""arXiv API client and feed helpers."""

from __future__ import annotations

import logging
import urllib.parse
import xml.etree.ElementTree as ET
from dataclasses import dataclass, field

from arxiv_copilot.schemas import ArxivPaper
from arxiv_copilot.utils.http import HttpClient

LOGGER = logging.getLogger(__name__)
ATOM = {"atom": "http://www.w3.org/2005/Atom", "arxiv": "http://arxiv.org/schemas/atom"}


@dataclass(slots=True)
class ArxivClient:
    http: HttpClient = field(default_factory=HttpClient)
    base_url: str = "https://export.arxiv.org/api/query"

    def fetch_by_id(self, arxiv_id: str) -> ArxivPaper:
        papers = self.search(f"id:{arxiv_id}", max_results=1)
        if not papers:
            raise LookupError(f"No arXiv paper found for {arxiv_id}")
        return papers[0]

    def newest(self, category: str = "cs.AI", *, max_results: int = 10) -> list[ArxivPaper]:
        return self.search(f"cat:{category}", max_results=max_results, sort_by="submittedDate", sort_order="descending")

    def search(self, query: str, *, max_results: int = 10, sort_by: str = "relevance", sort_order: str = "descending") -> list[ArxivPaper]:
        params = urllib.parse.urlencode(
            {
                "search_query": query,
                "start": 0,
                "max_results": max_results,
                "sortBy": sort_by,
                "sortOrder": sort_order,
            }
        )
        url = f"{self.base_url}?{params}"
        LOGGER.info("Fetching arXiv query: %s", query)
        return _parse_feed(self.http.get_text(url))


def _parse_feed(xml_text: str) -> list[ArxivPaper]:
    root = ET.fromstring(xml_text)
    papers: list[ArxivPaper] = []
    for entry in root.findall("atom:entry", ATOM):
        arxiv_id = _text(entry, "atom:id").rsplit("/", 1)[-1]
        pdf_url = None
        entry_url = None
        for link in entry.findall("atom:link", ATOM):
            href = link.attrib.get("href")
            if link.attrib.get("title") == "pdf" or link.attrib.get("type") == "application/pdf":
                pdf_url = href
            elif link.attrib.get("rel") == "alternate":
                entry_url = href
        categories = [category.attrib.get("term", "") for category in entry.findall("atom:category", ATOM)]
        papers.append(
            ArxivPaper(
                arxiv_id=arxiv_id,
                title=" ".join(_text(entry, "atom:title").split()),
                abstract=" ".join(_text(entry, "atom:summary").split()),
                authors=[_text(author, "atom:name") for author in entry.findall("atom:author", ATOM)],
                published=_text(entry, "atom:published") or None,
                updated=_text(entry, "atom:updated") or None,
                pdf_url=pdf_url,
                entry_url=entry_url,
                categories=[category for category in categories if category],
            )
        )
    return papers


def _text(element: ET.Element, path: str) -> str:
    found = element.find(path, ATOM)
    return found.text.strip() if found is not None and found.text else ""

File: artifacts/api-server/src/routes/arxiv.ts
import { Router, type IRouter } from "express";
import { XMLParser } from "fast-xml-parser";

const router: IRouter = Router();

const parser = new XMLParser({
  ignoreAttributes: false,
  attributeNamePrefix: "@_",
  isArray: (name) => ["entry", "author", "category", "link"].includes(name),
});

type ArxivEntry = {
  id: string;
  title: string;
  summary: string;
  author: Array<{ name: string }>;
  published: string;
  updated: string;
  category: Array<{ "@_term": string }>;
  link: Array<{ "@_href": string; "@_type"?: string; "@_rel"?: string }>;
};

function parseArxivId(rawId: string): string {
  return rawId.replace("http://arxiv.org/abs/", "").replace("https://arxiv.org/abs/", "").split("v")[0];
}

function toJson(entry: ArxivEntry) {
  const arxivId = parseArxivId(entry.id);
  const pdfLink = entry.link?.find((l) => l["@_type"] === "application/pdf");
  const absLink = entry.link?.find((l) => l["@_rel"] === "alternate");
  return {
    id: arxivId,
    arxiv_id: arxivId,
    title: entry.title?.trim().replace(/\s+/g, " ") ?? "",
    abstract: entry.summary?.trim().replace(/\s+/g, " ") ?? "",
    authors: (entry.author ?? []).map((a) => a.name).filter(Boolean),
    categories: (entry.category ?? []).map((c) => c["@_term"]).filter(Boolean),
    published: entry.published ?? "",
    updated: entry.updated ?? "",
    pdf_url: pdfLink?.["@_href"] ?? `https://arxiv.org/pdf/${arxivId}`,
    abs_url: absLink?.["@_href"] ?? `https://arxiv.org/abs/${arxivId}`,
  };
}

async function queryArxiv(params: Record<string, string>): Promise<ReturnType<typeof toJson>[]> {
  const qs = new URLSearchParams(params).toString();
  const res = await fetch(`http://export.arxiv.org/api/query?${qs}`, {
    headers: { "User-Agent": "ArxivCopilot/1.0 (replit.dev)" },
  });
  if (!res.ok) throw new Error(`arXiv API error ${res.status}`);
  const xml = await res.text();
  const parsed = parser.parse(xml);
  const entries: ArxivEntry[] = parsed?.feed?.entry ?? [];
  return entries.map(toJson);
}

router.get("/search", async (req, res) => {
  const q = (req.query.q as string) ?? "";
  const max = Math.min(Number(req.query.max) || 20, 50);
  if (!q) {
    res.status(400).json({ error: "q is required" });
    return;
  }
  const papers = await queryArxiv({
    search_query: `all:${q}`,
    max_results: String(max),
    sortBy: "relevance",
    sortOrder: "descending",
  });
  res.json(papers);
});

router.get("/category", async (req, res) => {
  const cat = (req.query.cat as string) ?? "cs.AI";
  const max = Math.min(Number(req.query.max) || 30, 50);
  const papers = await queryArxiv({
    search_query: `cat:${cat}`,
    max_results: String(max),
    sortBy: "submittedDate",
    sortOrder: "descending",
  });
  res.json(papers);
});

router.get("/paper", async (req, res) => {
  const id = (req.query.id as string) ?? "";
  if (!id) {
    res.status(400).json({ error: "id is required" });
    return;
  }
  const papers = await queryArxiv({ id_list: id, max_results: "1" });
  if (!papers[0]) {
    res.status(404).json({ error: "Paper not found" });
    return;
  }
  res.json(papers[0]);
});

router.get("/recent", async (req, res) => {
  const cats = (req.query.cats as string) ?? "cs.AI";
  const max = Math.min(Number(req.query.max) || 20, 50);
  const catList = cats.split(",").slice(0, 3);
  const query = catList.map((c) => `cat:${c.trim()}`).join(" OR ");
  const papers = await queryArxiv({
    search_query: query,
    max_results: String(max),
    sortBy: "submittedDate",
    sortOrder: "descending",
  });
  res.json(papers);
});

export default router;
